src/Symfony/Component/Security/Csrf/DoubleSubmitCsrfTokenManager.php

```diff
@@ -0,0 +1,268 @@
+ * send the X-Forwarded-* / Forwarded headers if you're behind one.
+ *
+ * Then, we validate the request using a cookie and a CsrfToken. If the cookie is found, it should
+ * contain the same value as the CsrfToken. A JavaScript snippet on the client side is responsible
+ * for performing this double-submission. The token value should be regenerated on every request
+ * using a cryptographically secure random generator.
+ *
+ * If either double-submit or Origin/Referer headers are missing, it typically indicates that
+ * JavaScript is disabled on the client side, or that the JavaScript snippet was not properly
+ * implemented, or that the Origin/Referer headers were filtered out.
+ *
+ * Requests lacking both double-submit and origin information are deemed insecure.
+ *
+ * When a session is found, a behavioral check is added to ensure that the validation method does not
+ * downgrade from double-submit to origin checks. This prevents attackers from exploiting potentially
+ * less secure validation methods once a more secure method has been confirmed as functional.
+ *
+ * On HTTPS connections, the cookie is prefixed with "__Host-" to prevent it from being forged on an
+ * HTTP channel. On the JS side, the cookie should be set with samesite=strict to strengthen the CSRF
+ * protection. The cookie is always cleared on the response to prevent any further use of the token.
+ *
+ * The $checkHeader argument allows the token to be checked in a header instead of or in addition to a
+ * cookie. This makes it harder for an attacker to forge a request, though it may also pose challenges
+ * when setting the header depending on the client-side framework in use.
+ *
+ * When a fallback CSRF token manager is provided, only tokens listed in the $tokenIds argument will be
+ * managed by this manager. All other tokens will be delegated to the fallback manager.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class DoubleSubmitCsrfTokenManager implements CsrfTokenManagerInterface
+{
+    public const TOKEN_MIN_LENGTH = 24;
+
+    public const CHECK_NO_HEADER = 0;
+    public const CHECK_HEADER = 1;
+    public const CHECK_ONLY_HEADER = 2;
+
+    /**
+     * @param self::CHECK_* $checkHeader
+     * @param string[]      $tokenIds
+     */
+    public function __construct(
+        private RequestStack $requestStack,
+        private ?LoggerInterface $logger = null,
+        private ?CsrfTokenManagerInterface $fallbackCsrfTokenManager = null,
+        private array $tokenIds = [],
+        private int $checkHeader = self::CHECK_NO_HEADER,
+        private string $cookieName = 'csrf-token',
+    ) {
+        if (!$cookieName) {
+            throw new \InvalidArgumentException('The cookie name cannot be empty.');
+        }
+
+        if (!preg_match('/^[-a-zA-Z0-9_]+$/D', $cookieName)) {
+            throw new \InvalidArgumentException('The cookie name contains invalid characters.');
```

Comment:
We can make this error more actionable by specifying the valid characters, e.g.
```suggestion
            throw new \InvalidArgumentException('The cookie name contains invalid characters, allowed characters: alfanumeric, -, _.');
```

Reply:
This is not supposed to happen, so not needed IMHO.